docs/Configuration.md
# Configuring PyCosworth

PyCosworth is a combination of code-reader/sensor reader, data logger and digital dashboard.

The software can be configured to enable or disable most of those options.

## Main Configuration File

Configuration for all aspects of the software are held in the file `libs/settings.py`.

The configuration file is divided up in to several sections:

* Enable or disable programme features
* Module configuration

---

## Enable Programme Features

**Main modules**

This section enables or disables the modules of the software which are activated at run-time. 

Note that all of the modules run in their own process and communicate their state, so whilst the GraphicsIO module is running, all of the other modules are not blocked from using the processor. The main reason is that the SensorIO module is *always* running in the background, gathering sensor readings, and is never blocked from running by one of the other modules.

*For this reason, it is highly reccomended that the software be run on a Raspberry Pi 3 which has 4 processor cores.*

 * USE_CONSOLE
    * Enable or disable the **ConsoleIO** text output module which outputs sensor information to the command prompt in a basic, text-mode display.
 * USE_BUTTONS
    * Enable the **GPIOButtionIO** joystick/button control module. Also enables keyboard key-presses to emulate joystick features. Without this module, no user control of the software is possible; it will just run as-configured.
* USE_GRAPHICS
    * Enable the **GraphicsIO** graphics output module. This module can display to various types of OLED hardware, either connected via the Raspberry Pi I2C or SPI bus. Enabling this module *also* activates an *emulated* display which is only possible on a machine with a working graphics display (for example a Pi with a desktop session, or a Linux desktop).
* USE_DATALOGGER
    * Enable the module which records sensor datastream to disk for later analysis.

**Customisation of modules**

Some of the module have additional major functionality that can be enabled or disabled - for example, activating certain types of additional sensors or display modes.

* USE_OLED_GRAPHICS
    * Within the **GraphicsIO** module, activate output to one or more OLED graphics display screens. These are mono, or grayscale, dot matrix style displays that are used to show sensor information, error status codes or visualise performance data in real time. Most development work has been done using the SSD1306 and SSH1102 monochrome displayed connected via a simple 4-wire I2C bus. You should read the documentation for the [Python luma.oled module](https://luma-oled.readthedocs.io/en/latest/hardware.html) to understand how to connect one of these devices to your computer/Pi.
* USE_SDL_GRAPHICS
    * Within the **GraphicsIO** module, activates an emulated OLED graphics display screen for each actual OLED screen that is configured. This way the visualisation and other display output can be tested on a desktop or laptop system without access to the specific OLED hardware. This requires a display that is supported by *SDL* libraries for graphics output. Generally this means X11 on Linux or Windows on PC.
* USE_COSWORTH
    * Try to activate Cosworth L8/P8 Pectel serial datastream sensors within the **SensorIO** module. See the [Pectel Serial Datastream](Pectel.md) guide which explains the format of the serial interface and how we use it. You will of course need the correct USB to RS232 adaptor on your laptop or Raspberry Pi to use it.
* USE_AEM
    * Try to activate AEM Wideband X-series AFR sensors within the **SensorIO** module. You need an AEM wideband sensor with serial datastream connected to a secondary USB to RS232 adaptor on your laptop or Raspberry Pi. The [AEM installation guide](https://www.aemelectronics.com/files/instructions/30-0300.pdf) shows the wiring to use for the serial data output, as well as the comms parameters.
* USE_SENSOR_DEMO
    * Activate demo-mode sensor data within the **SensorIO** module which supplies a never-ending stream of bogus, demo data, as if the software was really connected to active sensors which were sending data. If enabled, this can be activated or de-activated at will, from the software whilst running.

**Debug messages**

By default, only warnings and errors are output whilst the software is running. However, for debugging, either or both of these options can be set and will result in a lot more output from the software about what it is doing.

* INFO
    * Prints mainly status messages, some performance information
* DEBUG
    * Even more verbose output, including information of variable contents, timers, system calls, etc.
    
For most of the time, both items are best left as **False** in production mode.

---

## Module Configuration

### Sensor Information

* SENSOR_MAX_HISTORY
    * The number of previous readings to keep at any point in time for any sensor. This can be used to smooth readings, generate graphics/waveforms etc. Not currently used for the simple numeric display mode. **Reccomendation: 256**

* SENSOR_SLEEP_TIME
    * The amount of time, in seconds, to sleep between each loop through the entire list of sensors. This should be greater than 0 in order to prevent CPU lockup. Values of 0.02 to 0.05 would enable between 50 (1/0.02) and 20 (1/0.05) sensor passes per second. **Reccomendation: 0.02**

* SENSOR_ERROR_HEARTBEAT_TIMER
    * The time, in seconds, between error/status messages sent from the **SensorIO** process to the user interface. Lower values enable the user interface to react faster to errors connecting to the ECU and AFR sensors, but will reduce the overall responsiveness of the interface. **Reccomendation: 2-5 seconds**

### Cosworth ECU Settings

* COSWORTH_ECU_USB
    * The serial to USB device which is connected to the L8/P8/Pectel Datastream enabled ECU. The first USB to serial device on Linux will generally be **/dev/ttyUSB0**. There are no other settings to configure as the serial protocol is fixed.

### AEM Wideband AFR Settings

* AEM_USB
    * The serial to USB device which is connected to the AEM Wideband AFR sensor module/gauge.

### Graphics / OLED Display Settings

* GFX_MASTER_SIZE
    * The size of the emulated display and the physical OLED screen (if equipped). This is in the form of a Python tuple. PyCosworth has been mainly tested with 128x64 pixel resolution screens. Operation with other size displays is untested. **Reccomendation: (128, 64)**

* GFX_BOOT_LOGO
    * An image to be shown at power-on time. This should be a 1bpp image no larger than the size of your display configured at GFX_MASTER_SIZE. **Reccomendation: "logs/cosworth.bmp"**

* GFX_SLEEP_TIME
    * The time, in seconds, between updates of the emulated and physical screens. This is *not* the time between data updates - the screen could be refreshed faster than the data is updated, in which case the data will not have changed since the last update. Lower values will result in faster screen updates, but depending on the OLED screen being used may have tearing effects above a certain point. Again this should be left as a positive value to avoid unnecessary CPU use. Values of 0.01 (1/0.01 = 100 updates/sec) and 0.05 (1/0.02 = 20 updates/sec) should be tested to find the right one for your display device. **Reccomendation: 0.01 - 0.05**

* GFX_MASTER_WINDOW
    * Definitions for the main display window, including the list of SensorID's to show (and the sequence in which to show them), which OLED device to use (and which I2C address to connect to it with). This data structure should ideally be left as-is, other than to alter the sequence of sensors, if you desire a certain sensor to always be shown first.

* GFX_FONTS
    * Which fonts to use for various parts of the user interface.

### Datalogger Settings

* LOGGING_HEARTBEAT_TIMER
    * The time, in seconds, between update messages sent from the **DataLoggerIO** process to the user interface. This updates the user interface on whether the logger is running or not, or if it has encountered an error (such as disk space). **Reccomendation: 2-3 seconds**

* LOGGING_ACTIVE_SLEEP
    * Controls the granularity of the data logs which are written with sensor data. A higher value lowers the frequency at which a line of sensor data is written to disk. A value of **0.1** indicates that 10 lines of log data are written to disk in 1 second. The value of a sensor at the point it is written to disk is dependent on  the refresh value of that sensor. Bear in mind that with the Cosworth ECU particularly, we have a very slow serial baud rate that means we can only do, **at best** 60 single-byte sensor queries per second.
    * All sensors have a different refresh value (it doesn't make sense to look up battery voltage every 100 milliseconds, for example), and these are written in the back-end ECU support classes:
    * [iomodules/sensors/Cosworth.py](../iomodules/sensors/Cosworth.py)
    * [iomodules/sensors/AEM.py](../iomodules/sensors/AEM.py)

* LOGGING_SLEEP
    * How long to sleep, in seconds, before responding to an 'Activate Logging' signal. Again, a compromise between CPU over-use and responsiveness. **Reccomendation: 1-2 seconds**

* LOGGING_DIR
    * The directory where sensor logs should be written to. If the directory does not exist, it will be attempted to be created the first time the application runs. This should be in a location that the application can write to, not in a system directory or an administrative user.

* LOGGING_FILE_PREFIX
    * The first part of the filename of your sensor log files - this will be suffixed with an auto-incrementing numeric value. **Reccomendation: "pycosworth_"**

* LOGGING_FILE_SUFFIX
    * The last part of your sensor log file names. **Reccomendation: ".csv"**

libs/settings.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

# Settings.py - a configuration file for PyCosworth.
# Copyright (C) 2018  John Snowdon
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

##############################################################
#
# Weber / Magneti Marelli / Cosworth serial diagnostic port
#
##############################################################

# Total amount of time the main process sleeps until trying to load data from the
# receiving queue
MAIN_SLEEP_TIME = 0.02

# Maximum number of errors that can be retrieved and buffered
MAX_ERRORS = 255

# Standard types of message we may get back from the SensorIO process
TYPE_ERROR = "ERR_MSG"
TYPE_DATA = "DATA_MSG"
TYPE_STATUS = "STATUS_MSG"

##############################################################
#
# Turn various features on/off
#
##############################################################

# Selct the various worker modules
USE_MATRIX = False			# Output to a Matrix Orbital compatible character mode LCD
USE_CONSOLE = False			# Output to a standard terminal / command prompt
USE_BUTTONS = True			# Run a process which monitors Raspberry Pi GPIO buttons for button presses
USE_GRAPHICS = True			# Output to OLED modules or on-screen graphics
USE_DATALOGGER = True		# Run the datalogger module to record ecudata to disk
USE_OLED_GRAPHICS = True 	# Try to output to an OLED module
USE_SDL_GRAPHICS = True  	# Try to output to on-screen windows

# Sensor modules
USE_GEAR_INDICATOR = False	# Try to read gear indicator position via GPIO 
USE_COSWORTH = False 		# Try to connect to a Cosworth L8/P8 ECU over serial
USE_AEM = True				# Try to connect to an AEM Wideband AFR module over serial
USE_SENSOR_DEMO = False 	# Enable demo data mode from the SensorIO module instead of real data

# Should INFO category messages be shown
INFO = True
DEBUG = False

# Enable/Disable support for the Super Watchdog V2 Raspberry Pi hat
USE_PI_WATCHDOG = True

##############################################################
#
# Sensor information
#
##############################################################

# A list of the sensors we check on each normal loop in the SensorIO process.
#
# This also controls the refresh rate of the sensors. The Magneti Marelli hardware can refresh some data
# more frequently than others, but we can control, to some degree, how fast we refresh the data.
#
# We also have the possibility of adding other sensors to a Pi, or similar, that cannot be provided by an ECU
# for example, an analogue temperature sensor for outside air temp, a DIY gearstick position sensor, brake pedal angle, etc.
#
# Any sensor available in a back-end sensor module, needs to be listed below.
#
# 'sensorId' 		value is the keyword by which that sensor value is accessed in any display routine. DO NOT CHANGE!
# 'maxValue'		used to calculate maximum height of any graphs or displays for this sensor
# 'warnValue'		the value, which, if exceeded, the system will try to warn the driver of that sensor

# This needs to be simplified to remove min/max, unit, refresh etc. as it is all now defined per sensor module backend (see iomodules/sensors/Cosworth.py, demo.py etc.)
SENSORS = [
	{	'sensorId' 	: 'AFR',		'minValue' : 0,		'maxValue'	: 20,	'warnValue' : 15,	},
	{	'sensorId' 	: 'AMAL',	'minValue' : 0,		'maxValue'	: 100,	'warnValue' : 110,	},
	{	'sensorId' 	: 'BAT',		'minValue' : 0,		'maxValue'	: 14,	'warnValue' : 15,	},
	{	'sensorId' 	: 'CO',		'minValue' : 0,		'maxValue'	: 50,	'warnValue' : 100,	},	
	{	'sensorId' 	: 'ECT',		'minValue' : 0,		'maxValue'	: 150,	'warnValue' : 110, 	},
	{	'sensorId'	: 'IAT',		'minValue' : 0,		'maxValue'	: 60,	'warnValue' : 50,	},
	{	'sensorId' 	: 'IGNADV',	'minValue' : 0,		'maxValue'	: 40,	'warnValue' : 36,	},
	{	'sensorId' 	: 'INJDUR',	'minValue' : 0,		'maxValue'	: 5,		'warnValue' : 20,	},	
	{	'sensorId' 	: 'MAP',		'minValue' : -350,	'maxValue'	: 3000,	'warnValue' : 2500,	},
	{ 	'sensorId'	: 'RPM',		'minValue' : 0,		'maxValue'	: 7500, 'warnValue' : 6000,	},
	{	'sensorId' 	: 'TPS',		'minValue' : -0.3,	'maxValue'	: 90,	'warnValue' : 100,	},
]
# A list of all sensor id's
SENSOR_IDS = []
for s in SENSORS:
	SENSOR_IDS.append(s['sensorId'])
	
# How many previous sensor samples, for each sensor, to keep in memory
SENSOR_MAX_HISTORY = 256

# The amount of time, in seconds, that the SensorIO process sleeps between each loop
SENSOR_SLEEP_TIME = 0.05

# How often to sleep between broadcasting ECU/comms error messages
SENSOR_ERROR_HEARTBEAT_TIMER = 5

##########################################################
#
# Cosworth ECU settings
#
##########################################################
COSWORTH_ECU_TYPE = "L8 Pectel" # see iomodules/sensors/CosworthSensors.py for available types

# Which USB interface your USB to serial device is on
COSWORTH_ECU_USB = "/dev/ttyUSB0" 
	
#########################################################
#
# AEM Wideband AFR settings
#
#########################################################
AEM_USB = "/dev/ttyUSB1"
	
#######################################################
#
# Demo data module config
#
#######################################################

# The number of steps between the minValue and the maxValue of a
# sensor to simulate.
DEMO_STEPS = 64

#######################################################
#
# GPIO module config - this sends button presses and 
# configuration change messages between the various worker
# processes.
#
#######################################################

# How long to sleep between button scans
GPIO_SLEEP_TIME = 0.2

# Length of time each press takes (min time, max time)
BUTTON_TIME_SHORT 	= (0, 	0.3)
BUTTON_TIME_MEDIUM 	= (0.5, 1.0)
BUTTON_TIME_LONG 	= (2.0, 5.0)

# Button/status definitions
BUTTON_TOGGLE_DEMO 			= "d" # Start/stop demo mode
BUTTON_LOGGING_RUNNING 		= "L" # Logging is running
BUTTON_LOGGING_STOPPED 		= "l" # Logging is stopped
STATUS_ECU_ERROR		 		= "e1" # Main ECU error
STATUS_AEM_ERROR		 		= "e2" # AEM module error
STATUS_POW_ERROR		 		= "e3" # AEM module error
STATUS_ECU_OK	 			= "o1" # Main ECU error
STATUS_AEM_OK		 		= "o2" # AEM module error
STATUS_POW_OK		 		= "o3" # AEM module error
STATUS_SHUTDOWN				= "xxx"
STATUS_DEMO_ENABLED	 		= "d1" #
STATUS_DEMO_DISABLED	 		= "d2" # 
BUTTON_LOGGING_STATUS 		= "S" # Logging status/heartbeat response
BUTTON_RESET_COSWORTH_ECU 	= "R" # Reset Cosworth ECU serial comms
BUTTON_RESET_AEM_ECU 		= "r" # Reset AEM serial comms

# For the simple, 3 button interface
BUTTON_LOGGING_TOGGLE 		= "1" # Logging is stopped or started
BUTTON_SENSOR_NEXT			= "2" # Select next sensor
BUTTON_RESET_ECU				= "3" # Reset all comms

# Button message types
MESSAGE_TYPE_PRESS 	= 0x01 # message is a button press
MESSAGE_TYPE_PAUSE 	= 0xFE # message is to pause logging/display
MESSAGE_TYPE_EXIT 	= 0xFF # message is to shut down

# Button durations
BUTTON_SHORT 	= 0x11 # short duration press
BUTTON_MEDIUM 	= 0x12 # medium duration press
BUTTON_LONG 	= 0x13 # long press

# Set destination types
BUTTON_DEST_ALL 		= 0x00 # send to all workers
BUTTON_DEST_MAIN 		= 0x01 # send to main PyCosworth process
BUTTON_DEST_SENSORIO 	= 0x02 # send to serial process
BUTTON_DEST_CONSOLEIO 	= 0x03 # send to console process
BUTTON_DEST_MATRIXIO 	= 0x04 # send to matrix lcd process
BUTTON_DEST_GRAPHICSIO 	= 0x05 # send to SDL/OLED graphics process
BUTTON_DEST_DATALOGGER 	= 0x06 # send to datalogger process

# Mapping of buttons to modules
# i.e. button 1 and 2 to GraphicsIO, button 3 to datalogger, etc
BUTTON_MAP = {
	BUTTON_RESET_ECU				: { 'dest' : BUTTON_DEST_ALL }, # Attempt comms reset
	BUTTON_TOGGLE_DEMO			: { 'dest' : BUTTON_DEST_SENSORIO }, # Toggle demo start/stop
	BUTTON_LOGGING_TOGGLE		: { 'dest' : BUTTON_DEST_DATALOGGER }, # Toggle demo start/stop
	BUTTON_SENSOR_NEXT			: { 'dest' : BUTTON_DEST_GRAPHICSIO }, # Show next sensor
	BUTTON_LOGGING_RUNNING		: { 'dest' : BUTTON_DEST_ALL }, # Logging is running
	BUTTON_LOGGING_STOPPED		: { 'dest' : BUTTON_DEST_ALL }, # Logging is stopped
	STATUS_ECU_ERROR				: { 'dest' : BUTTON_DEST_GRAPHICSIO }, # ECU comms problem
	STATUS_AEM_ERROR				: { 'dest' : BUTTON_DEST_GRAPHICSIO }, # AEM AFR comms problem
	STATUS_ECU_OK				: { 'dest' : BUTTON_DEST_GRAPHICSIO }, # ECU comms problem
	STATUS_AEM_OK				: { 'dest' : BUTTON_DEST_GRAPHICSIO }, # AEM AFR comms problem
}

#######################################################
#
# Graphics module config
#
#######################################################

# Size of the main control screen
# This is the master screen that shows settings, config
# and higher resolution data
# NOTE: The on-desktop SDL window will be created at the same size
GFX_MASTER_SIZE = (128, 64)

# Size of the OLED miniature screen.
# These are the sub-screens that generally display a single gauge or sensor.
# NOTE: The on-desktop SDL window will be created at the same size
GFX_OLED_SIZE = GFX_MASTER_SIZE

# Where we keep images, icons etc.
GFX_ASSETS_DIR = "images/"
GFX_CACHE_DIR = "cache/"

# Boot up logo for the OLED screens
GFX_BOOT_LOGO 		= "logo/ford.bmp"
GFX_BOOT_LOGO1 		= "logo/cosworth.bmp"
GFX_BOOT_LOGO2 		= "logo/pycosworth.bmp"
GFX_BOOT_LOGO_BIG 	= "logo/cosworth_outline.bmp"

# Time to sleep between graphics updates
# The lower the sleep time, the smoother the gfx display,
# but the more CPU power the process will take.
#
# NOTE: Certain I2C connected OLED devices are effectively
# throttled in how fast they can update, so decreasing this
# value may not result in a faster refresh for those devices.
# Enable 'INFO' logging, above and check the output messages from 
# the console to see what the average fps for rendering to
# the OLED devices is;
#
# For example, with a single, no-name SH1106 controller, 128x64 
# pixel OLED on I2C bus, I see 17-20fps with GFX_SLEEP_TIME = 0.05, 
# but decreasing the sleep time doesn't result in any faster refresh.
#
# Assuming your hardware can keep up, 0.05 should average 20fps, which
# is plenty fast to update an LCD/OLED gauge.
GFX_SLEEP_TIME = 0.01

# How long a period to measure framerate over
GFX_FRAME_COUNT_TIME = 5

GFX_MODE_NUMERIC = "Numeric"
GFX_MODE_OFF = "OFF"

# Available modes that a sensor can be shown in
#GFX_MODES = [GFX_MODE_WAVEFORM, GFX_MODE_SEGMENTS, GFX_MODE_CLOCK, GFX_MODE_LINE, GFX_MODE_OFF]
GFX_MODES = [GFX_MODE_NUMERIC]

GFX_MASTER_WINDOW = {
	'windowName'			: 'Master',
	'oledType'			: 'sh1106',
	'width'				: GFX_MASTER_SIZE[0],
	'height'				: GFX_MASTER_SIZE[1],
	'spiAddress'			: 0,
	'value_refreshTime'	: 0.05,
	'sdl_framebuffer'	: None,
	'luma_framebuffer'	: None,
	'luma_driver'		: None,
	'screen_refreshTime': 0.02,
	'i2cPort'			: 8,
	'i2cAddress'		: 0x3c,
	'mode'				: [GFX_MODE_NUMERIC],
	'currentModeIdx'		: 0,
	'currentMode'		: GFX_MODE_NUMERIC,
	'sensorIds'			: ['AFR', 'AMAL', 'BAT', 'CO', 'ECT', 'IAT', 'IGNADV', 'INJDUR', 'MAP', 'RPM', 'TPS'],
	'currentSensorIdx'	: 0,
	'screen_cycleTime'	: 5,
	'value_refreshTime'	: 0.1,
	'screen_refreshTime': 0.05,
}

# All of the fonts used in the graphics routines
#
# Fonts available from: https://www.dafont.com/
GFX_FONTS = {
	'sans' 	: { 
		'plain' 	: { 'font' : GFX_ASSETS_DIR + 'fonts/DejaVuSansCondensed.ttf' },
		'italic' 	: { 'font' : GFX_ASSETS_DIR + 'fonts/DejaVuSansCondensed-Oblique.ttf' },
		'bold' 		: { 'font' : GFX_ASSETS_DIR + 'fonts/DejaVuSansCondensed-Bold.ttf' },
		'bolditalic': { 'font' : GFX_ASSETS_DIR + 'fonts/DejaVuSansCondensed-BoldOblique.ttf' },
	},
	'pixel'	: {
		'plain' 	: { 'font' : GFX_ASSETS_DIR + 'fonts/pixelmix_8px.ttf' },
		'large' 	: { 'font' : GFX_ASSETS_DIR + 'fonts/Minecraft_16px.ttf' },
		'header' 	: { 'font' : GFX_ASSETS_DIR + 'fonts/neoletters_16px.ttf' },
	},
	'lcd'	: {
		'plain' 	: { 'font' : GFX_ASSETS_DIR + 'fonts/CFLCD-Regular.ttf' },
	},
}


#######################################################
#
# Logging module config
#
#######################################################

# Broadcast logging status every 'X' seconds
LOGGING_HEARTBEAT_TIMER = 3

# How often to sleep between loops, should be no more than the sensor module
# otherwise we may miss datapoints
LOGGING_ACTIVE_SLEEP = 0.1 	# How long to sleep while recording (interval between data recorded to log file)
LOGGING_SLEEP = 1 			# How long to sleep while recording inactive

LOGGING_DIR = "logs"
LOGGING_FILE_PREFIX = "pycosworth_"
LOGGING_FILE_SUFFIX = ".csv"

########################################################
#
# Watchdog timers and Power Monitoring
# 
########################################################

# How often, in seconds, to send a 'alive' message to the 
# watchdog timer on the Raspberry Pi UPS hat
WATCHDOG_HEARTBEAT_TIMER = 30

# How often, in seconds, to detect the current power state
# of the Raspberry Pi
WATCHDOG_POWER_TIMER = 5

# How long to run the Pi on battery backup after a power-loss is
# detected, until we begin a controlled shutdown
WATCHDOG_POWER_SHUTDOWN_TIMER = 20

# How many volts should be nominal input
WATCHDOG_POWER_NOMINAL = 5.0